cores/esp8266/heap.cpp

```diff
@@ -23,370 +126,540 @@ extern "C" void z2EapFree(void *ptr, const char* file, int line) __attribute__((
 
 extern "C" {
 
-#if defined(UMM_POISON_CHECK) || defined(UMM_POISON_CHECK_LITE)
-#define UMM_MALLOC(s)           umm_poison_malloc(s)
-#define UMM_CALLOC(n,s)         umm_poison_calloc(n,s)
-#define UMM_REALLOC_FL(p,s,f,l) umm_poison_realloc_fl(p,s,f,l)
-#define UMM_FREE_FL(p,f,l)      umm_poison_free_fl(p,f,l)
-#define STATIC_ALWAYS_INLINE
+///////////////////////////////////////////////////////////////////////////////
+// Select from various heap function renames that facilitate inserting debug
+// wrappers, and tradditional names for the non-debug case.
+//
+
+/*
+ * With any debug options listed above, umm_malloc changes its heap API names
+ * from malloc, calloc, realloc, and free to umm_malloc, umm_calloc,
+ * umm_realloc,  and umm_free.
+ *
+ */
+#undef STATIC_ALWAYS_INLINE
+#undef ENABLE_THICK_DEBUG_WRAPPERS
+
+#if defined(UMM_POISON_CHECK_LITE)
+/*
+ * umm_malloc will build with umm_poison_* wrappers for each Heap API.
+ *
+ * Support debug wrappers that need to include handling poison
+ */
+#define UMM_MALLOC_FL(s,f,l,c)    umm_poison_malloc(s)
+#define UMM_CALLOC_FL(n,s,f,l,c)  umm_poison_calloc(n,s)
+#define UMM_REALLOC_FL(p,s,f,l,c) umm_poison_realloc_flc(p,s,f,l,c)
+#define UMM_FREE_FL(p,f,l,c)      umm_poison_free_flc(p,f,l,c)
```

Comment:
Nice! That would be convenient.

Thoughts and Concerns:
1) Can `source_location` be used from `.c` modules?
2) Will function names be stored in `PROGMEM` or DRAM?
3) Their example shows a longer (full definition) string than we have used for the function name, which could cause build size issues for debug builds.

Reply:
Hmm, I am not able to build with `#include <source_location>`
```
Detecting libraries used...
/home/mhightow/Arduino/hardware/esp8266com/esp8266/tools/xtensa-lx106-elf/bin/xtensa-lx106-elf-g++ -D__ets__ -DICACHE_FLASH -U__STRICT_ANSI__ -D_GNU_SOURCE -DESP8266 -Os @/tmp/arduino_build_311930/core/build.opt -I/home/mhightow/Arduino/hardware/esp8266com/esp8266/tools/sdk/include -I/home/mhightow/Arduino/hardware/esp8266com/esp8266/tools/sdk/lwip2/include -I/home/mhightow/Arduino/hardware/esp8266com/esp8266/tools/sdk/libc/xtensa-lx106-elf/include -I/tmp/arduino_build_311930/core -c @/home/mhightow/Arduino/hardware/esp8266com/esp8266/tools/warnings/none-cppflags -g -free -fipa-pta -Werror=return-type -mlongcalls -mtext-section-literals -fno-rtti -falign-functions=4 -std=gnu++17 -ffunction-sections -fdata-sections -fexceptions -DMMU_IRAM_SIZE=0xC000 -DMMU_ICACHE_SIZE=0x4000 -DFP_IN_IROM -w -x c++ -E -CC -DNONOSDK22x_190703=1 -DF_CPU=80000000L -DLWIP_OPEN_SRC -DTCP_MSS=536 -DLWIP_FEATURES=1 -DLWIP_IPV6=0 -DDEBUG_ESP_PORT=Serial -DARDUINO=10819 -DARDUINO_ESP8266_NODEMCU_ESP12E -DARDUINO_ARCH_ESP8266 "-DARDUINO_BOARD=\"ESP8266_NODEMCU_ESP12E\"" "-DARDUINO_BOARD_ID=\"nodemcuv2\"" -DLED_BUILTIN=2 -DFLASHMODE_DIO -I/home/mhightow/Arduino/hardware/esp8266com/esp8266/cores/esp8266 -I/home/mhightow/Arduino/hardware/esp8266com/esp8266/variants/nodemcu /tmp/arduino_build_311930/sketch/SrcLoc.ino.cpp -o /dev/null
SrcLoc:3:10: fatal error: source_location: No such file or directoryAlternatives for source_location: []

    3 | #include <source_location>
      |          ^~~~~~~~~~~~~~~~~
compilation terminated.
```
I also noticed that path `... esp8266com/esp8266/tools/sdk/libc/xtensa-lx106-elf/include` does not exist on my machine, even after rerunning `./get.py`

Reply:
Not available in C, as it would need default-arg constructed by the caller.
Not available in -std=c++17, part of the -std=c++2a

What I meant was copying the behaviour, at least in structure which is then passed to the func. May be significantly larger in size, on second though :/
i.e. `static struct foo { .file = ..., .lineno = ...}; foo.caller = __builtin__(); func(..., &foo);`

Plus, we do have a hack around `__FILE__` which is not applied to source\_location to put strings in progmem as default `.rodata` is still in ram